python/lsst/pipe/base/graph/_loadHelpers.py

```diff
@@ -0,0 +1,310 @@
+        quantumToNodeId: Dict[Quantum, NodeId] = {}
+        loadedTaskDef = {}
+        # loop over the nodes specified above
+        for node in nodes:
+            # Get the bytes to read from the map
+            start, stop = self.map[node]
+            start += self.headerSize
+            stop += self.headerSize
+
+            # read the specified bytes, will be overloaded by subclasses
+            # bytes are compressed, so decompress them
+            dump = lzma.decompress(_readBytes(start, stop))
+
+            # reconstruct node
+            qNode = pickle.loads(dump)
+
+            # read the saved node, name. If it has been loaded, attach it, if
+            # not read in the taskDef first, and then load it
+            nodeTask = qNode.taskDef
+            if nodeTask not in loadedTaskDef:
+                # Get the byte ranges corresponding to this taskDef
+                start, stop = self.taskDefMap[nodeTask]
+                start += self.headerSize
+                stop += self.headerSize
+
+                # load the taskDef, this method call will be overloaded by
+                # subclasses.
+                # bytes are compressed, so decompress them
+                taskDef = pickle.loads(lzma.decompress(_readBytes(start, stop)))
+                loadedTaskDef[nodeTask] = taskDef
+            # Explicitly overload the "frozen-ness" of nodes to attach the
+            # taskDef back into the un-persisted node
+            object.__setattr__(qNode, 'taskDef', loadedTaskDef[nodeTask])
+            quanta[qNode.taskDef].add(qNode.quantum)
+
+            # record the node for later processing
+            quantumToNodeId[qNode.quantum] = qNode.nodeId
+
+        # Get the id of this graph
+        _buildId = qNode.nodeId.buildId if nodes else None  # type: ignore
+
+        # construct an empty new QuantumGraph object, and run the associated
+        # creation method with the un-persisted data
+        qGraph = object.__new__(QuantumGraph)
+        qGraph._buildGraphs(quanta, _quantumToNodeId=quantumToNodeId, _buildId=_buildId)
+        return qGraph
+
+    def _readBytes(self, start: int, stop: int) -> bytes:
+        """Loads the specified byte range from the ButlerURI object
+
+        In the base class, this actually will read all the bytes into a buffer
+        from the specified ButlerURI object. Then for each method call will
+        return the requested byte range. This is the most flexible
+        implementation, as no special read is required. This will not give a
+        speed up with any sub graph reads though.
+        """
+        if not hasattr(self, 'buffer'):
+            self.buffer = self.uriObject.read()
+        return self.buffer[start:stop]
+
+    def close(self):
+        """Cleans up an instance if needed. Base class does nothing
+        """
+        pass
+
+
+@register_helper(ButlerS3URI)
+class S3LoadHelper(DefaultLoadHelper):
```

Comment:
Not reviewing, but I'd rather we added byte range support to `ButlerURI.read()` directly -- it seems generically useful and if a URI scheme can't support it I guess we have to decided whether to complain or download the full file and do local read.  WebDAV is missing from this so I'm not sure if that means you didn't look at it or it can't support it.

Reply:
I just have not done webdav yet, I was going to tackle it in another follow-up ticket so as to unblock the ticket andy is working on.

I agree that if possible it would be nice to get this back upstream, as a comment in the file says. However, it will fundamentally change ButlerURI, and I was unprepared to do that with you away, or without proper planning. So I wrote it is such a way that future updates to ButlerURI would not affect the public api.

Basically ButlerURI supports the api of a file handle, but doesn't keep internal state on the position of the stream (for instance no seek method) so it does not behave as one. This is evident in the File uri where a open call happens for every read, and always assume reading bytes happens from the start of the file. Changing this changes the expectation of how the objects can be used (and or potentially means keeping handles open and forcing ButlerURI to be used as a context manager (which I know it is in some ways already))

Im happy if we want to make those changes though.

Reply:
Ok, I misunderstood what you had done. I was more than happy for `ButlerURI.read()` to take an offset parameter. I'd need to think more about making it something that looks like an open filehandle that remembers seek position.

Reply:
I am happy to write something related to expected behavior, trade offs etc.

And maybe you are right that an offset parameter is fine, but it will make a potentially confusing interface that is close to the io module but behaves differently. But it's something we should think through.

I also should have said above, I didn't specialize it for webdav, but it will work fine. It just won't have the benefit of loading select chunks, it will use the default loader and basically load the whole thing and subset, which is not exactly ideal from a network perspective. I don't think it will be hard, and maybe I will have time to do it while andy is reviewing, if not I'll file a follow up ticket.